test/integration/master/kube_apiserver_test.go

```diff
@@ -85,6 +86,55 @@ func TestRun(t *testing.T) {
 	}
 }
 
+func endpointReturnsStatusOK(client *kubernetes.Clientset, path string) bool {
+	var status int
+	res := client.CoreV1().RESTClient().Get().AbsPath(path).Do()
+	status = 0
+	res.StatusCode(&status)
+	return status == http.StatusOK
+}
+
+func TestDelayedHealthzAndReadyz(t *testing.T) {
+	hc := &delayedCheck{}
+	instanceOptions := &kubeapiservertesting.TestServerInstanceOptions{
+		InjectedHealthzChecker: hc,
+	}
+	server := kubeapiservertesting.StartTestServerOrDie(t, instanceOptions, []string{"--post-start-hook-grace-period", "5s"}, framework.SharedEtcd())
+	defer server.TearDownFn()
+
+	client, err := kubernetes.NewForConfig(server.ClientConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	if endpointReturnsStatusOK(client, "/readyz") {
+		t.Fatalf("readyz should start unready")
+	}
+	// we need to wait longer than our grace period
+	err = wait.Poll(100*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
+		if !endpointReturnsStatusOK(client, "/healthz") {
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("healthz should have become unhealthy: %v", err)
+	}
+	hc.makeHealthy()
+	err = wait.Poll(100*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
+		if endpointReturnsStatusOK(client, "/healthz") {
```

Comment:
return endpointReturnsStatusOK(client, "/healthz"), nil

Reply:
oh, that's nice :)